src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1603,6 +1613,49 @@ public override BoundNode VisitCall(BoundCall node)
             return null;
         }
 
+        /// <summary>
+        /// For each argument, figure out if its corresponding parameter is annotated with NotNullWhenFalse or
+        /// EnsuresNotNull.
+        /// </summary>
+        private static ImmutableArray<AttributeAnnotations> GetAnnotations(int numArguments,
+            bool expanded, ImmutableArray<ParameterSymbol> parameters, ImmutableArray<int> argsToParamsOpt)
+        {
+            ArrayBuilder<AttributeAnnotations> builder = null;
+
+            for (int i = 0; i < numArguments; i++)
+            {
+                (ParameterSymbol parameter, _) = GetCorrespondingParameter(i, parameters, argsToParamsOpt, expanded);
+                AttributeAnnotations annotations = parameter?.FlowAnalysisAnnotations ?? AttributeAnnotations.None;
+
+                // We'll ignore NotNullWhenFalse that is misused in metadata
+                if ((annotations & AttributeAnnotations.NotNullWhenFalse) != 0 &&
+                    parameter.ContainingSymbol.GetTypeOrReturnType().SpecialType != SpecialType.System_Boolean)
+                {
+                    annotations &= ~AttributeAnnotations.NotNullWhenFalse;
+                }
+
+                // We'll ignore EnsuresNotNull that is misused in metadata
+                if ((annotations & AttributeAnnotations.EnsuresNotNull) != 0 &&
+                    (parameter.Type?.IsValueType != false || parameter.IsParams))
```

Comment:
>parameter [](start = 21, length = 9)

Should we use original definition for this check? Is generic case covered by tests?

Reply:
I'll add tests.
To do that more easily, I'll pull the change to remove the warning for misused attributes in this PR. That'll also address some other feedback.

Reply:
Added tests `EnsuresNotNull_Generic_WithRefType`, `EnsuresNotNull_Generic_WithValueType` and `EnsuresNotNull_Generic_WithUnconstrainedGenericType`.

---
In reply to: [187948601](https://github.com/dotnet/roslyn/pull/26656#discussion_r187948601) [](ancestors = 187948601)